dask/dataframe/multi.py

```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
+        frames.append(next)
+
+    return pd.merge_asof(left, pd.concat(frames), **kwargs)
+
+
+def merge_asof_indexed(left, right, **kwargs):
+    left = fix_overlap(left)
+    right = fix_overlap(right)
+
+    dsk = dict()
+    name = 'asof-join-indexed-' + tokenize(left, right, **kwargs)
+    meta = pd.merge_asof(left._meta_nonempty, right._meta_nonempty, **kwargs)
+
+    dependencies = [left, right]
+    tails = heads = None
+    if kwargs['direction'] in ['backward', 'nearest']:
+        tails = compute_tails(right, by=kwargs['right_by'])
+        dependencies.append(tails)
+    if kwargs['direction'] in ['forward', 'nearest']:
+        heads = compute_heads(right, by=kwargs['right_by'])
+        dependencies.append(heads)
+
+    for i, J in enumerate(pair_partitions(left.divisions, right.divisions)):
+        frames = []
+        for j, lower, upper in J:
+            slice = (methods.boundary_slice, (left._name, i),
+                     lower, upper, False)
+            tail = (tails._name, j) if tails is not None else None
+            head = (heads._name, j) if heads is not None else None
+            frames.append((apply, merge_asof_padded, [slice, (right._name, j),
+                           tail, head], kwargs))
+        dsk[(name, i)] = (pd.concat, frames)
+
+    graph = HighLevelGraph.from_collections(name, dsk,
+                                            dependencies=dependencies)
+    return new_dd_object(graph, name, meta, left.divisions)
+
+
+@wraps(pd.merge_asof)
+def merge_asof(left, right, on=None, left_on=None, right_on=None,
+               left_index=False, right_index=False, by=None, left_by=None,
+               right_by=None, suffixes=('_x', '_y'), tolerance=None,
+               allow_exact_matches=True, direction='backward'):
+    if direction not in ['backward', 'forward', 'nearest']:
+        raise ValueError("Invalid merge_asof direction. Choose from 'backward',\
+              'forward', or 'nearest'")
+
+    kwargs = {'on': on, 'left_on': left_on, 'right_on': right_on,
+              'left_index': left_index, 'right_index': right_index,
+              'by': by, 'left_by': left_by, 'right_by': right_by,
+              'suffixes': suffixes, 'tolerance': tolerance,
+              'allow_exact_matches': allow_exact_matches,
+              'direction': direction}
+
+    if left is None or right is None:
+        raise ValueError("Cannot merge_asof on empty DataFrames")
+
+    if isinstance(left, pd.DataFrame) and isinstance(right, pd.DataFrame):
+        return pd.merge_asof(left, right, **kwargs)
+
+    if on is not None:
+        left_on = right_on = on
+    for o in [left_on, right_on]:
+        if isinstance(o, _Frame):
+            raise NotImplementedError(
+                "Dask collections not currently allowed in merge columns")
+
+    if not is_dask_collection(left):
+        left = from_pandas(left, npartitions=1)
+    if left_on is not None:
+        left = left.set_index(left_on, sorted=True)
```

Comment:
`left` and `right` are coming straight from the user, right? Saying `sorted=True` when it's not in fact sorted is dangerous. I'm not sure we should do that here.

Reply:
It sure is dangerous. However, in the documentation for merge_asof it says the input data must be sorted. I don't see any good way of verifying that it's sorted before doing this operation without messing up the complexity, so I would just be inclined to trust the user. https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.merge_asof.html

Reply:
Pandas validates the sorting though, 

```python
In [7]: >>> pd.merge_asof(left, right[::-1], on='a')
   ...:
   ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-481523b2ecc3> in <module>
----> 1 pd.merge_asof(left, right[::-1], on='a')
      2

~/sandbox/pandas/pandas/core/reshape/merge.py in merge_asof(left, right, on, left_on, right_on, left_index, right_index, by, left_by, right_by, suffixes, tolerance, allow_exact_matches, direction)
    464                     allow_exact_matches=allow_exact_matches,
    465                     direction=direction)
--> 466     return op.get_result()
    467
    468

~/sandbox/pandas/pandas/core/reshape/merge.py in get_result(self)
   1295
   1296     def get_result(self):
-> 1297         join_index, left_indexer, right_indexer = self._get_join_info()
   1298
   1299         # this is a bit kludgy

~/sandbox/pandas/pandas/core/reshape/merge.py in _get_join_info(self)
    758         else:
    759             (left_indexer,
--> 760              right_indexer) = self._get_join_indexers()
    761
    762             if self.right_index:

~/sandbox/pandas/pandas/core/reshape/merge.py in _get_join_indexers(self)
   1535                 raise ValueError(msg_missings.format(side='right'))
   1536             else:
-> 1537                 raise ValueError(msg_sorted.format(side='right'))
   1538
   1539         # initial type conversion as needed

ValueError: right keys must be sorted
```

Can you add a test ensuring that `dd.merge_asof` raises in a similar case for dask arrays?

Reply:
I'm not sure we can reasonably verify this without doing a full-pass computation.

Reply:
@TomAugspurger what do you think we should do here?  I suspect that verifying sortedness is hard to do lazily.

Reply:
Since pandas raises, I think we’re ok. Would be good to add a test to verify.

Reply:
To be clear, pandas would be the one to raise when we execute the graph.

Reply:
Pandas wouldn't raise if the out-of-order values cross the partitions. If we _must_ check this then we could always `map_partitions` `min` and `max` of the `by` column to each partition and check the divisions. But I would rather leave the burden on the user to ensure that their data is ordered then mess up our runtime.